### PR TITLE
[ignition-cmake2/ignition-modularscripts/ignition-common3] Add options, support pkgconfig and fix usage for common3-graph

### PR DESCRIPTION
**Pull request by JackBoosY**
Fix usage when using `find_package(ignition-common3-graphics CONFIG REQUIRED)`:
```
CMake Error at /home/work/vcpkg/scripts/buildsystems/vcpkg.cmake:573 (_add_executable):
  Target "sample" links to target "GTS::GTS" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:45 (add_executable)


CMake Error at /home/work/vcpkg/scripts/buildsystems/vcpkg.cmake:573 (_add_executable):
  Target "sample" links to target "FreeImage::FreeImage" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:45 (add_executable)


CMake Error at /home/work/vcpkg/scripts/buildsystems/vcpkg.cmake:573 (_add_executable):
  Target "sample" links to target "TINYXML2::TINYXML2" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:45 (add_executable)
```

Related: https://github.com/microsoft/vcpkg/pull/24969

Already tested this usage in x64-linux.

### CODE REVIEW

File: ports/ignition-cmake2/portfile.cmake [hunk 1]
```diff
@@ -5,5 +5,9 @@ ignition_modular_library(NAME cmake
                          SHA512 e39ed44ae6f7ccc338412f466f1257f88989e0818bee801ddbe09350e906cd9ce709be24356310fdbfde22d1b5b5846fed0aa794c06dcf7caf82748a07b428d6
                          PATCHES FindGTS.patch)
 
+# For ignition-common3
+file(INSTALL "${SOURCE_PATH}/cmake/FindGTS.cmake" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
```
Comment (JackBoosY) [inline ports/ignition-cmake2/portfile.cmake:9]: ignition uses internal FindGTS.cmake and GTS doesn't export the cmake configurations.
All the internal find module are stored in the cmake2 port source.

File: ports/ignition-common3/fix-dependencies.patch [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -52,9 +52,9 @@ set(IGN_MATH_VER ${ignition-math6_VERSION_MAJOR})
+ #--------------------------------------
+ # Find Tinyxml2
+ if(USE_EXTERNAL_TINYXML2)
+-  ign_find_package(TINYXML2 PRETTY tinyxml2
++  ign_find_package(tinyxml2 EXTRA_ARGS CONFIG
+     REQUIRED_BY graphics
+-    PRIVATE_FOR graphics)
```
[the remark this answers is not in the document]
Comment (JackBoosY) [reply ports/ignition-common3/fix-dependencies.patch:12]: Using `PRIVATE_FOR` will not write find_dependency to _config.cmake_, however cmake will still export these private targets to `INTERFACE_LINK_LIBRARIES`.

**Review by LilyWangLL — APPROVED** [review]

**Comment by talregev** [comment]
Thank you for the fix

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
First, we should wait for https://github.com/microsoft/vcpkg/pull/25034 to merge which adds a standardized method for consuming GTS (pkg-config).

Then,
1. `FindGTS.cmake` should be replaced/implemented via the pkgconfig mechanism above
2. `ign_find_package` should probably be patched to make `PRIVATE_FOR` not suppress `find_dependency`
3. We should not need a `vcpkg-cmake-wrapper` here. Any changes needed should be written into the `-config.cmake` created by the port.

**Comment by JackBoosY** [comment]
@ras0219-msft  The official doc said:
```
# [PRIVATE_FOR]: Optional. If provided, the list that follows it must indicate
#                which library components depend on this package privately (i.e.
#                the package should not be included in its list of interface
#                libraries). This is only relevant for components that follow
#                the REQUIRED_BY command. Note that the PRIVATE argument does
#                not apply to components specified by REQUIRED_BY. This argument
#                MUST be given for components whose private dependencies have
#                been specified with REQUIRED_BY.
```
This parameter **only** affects `cmake-config.cmake` and `pkgconfig.pc`:
https://github.com/gazebosim/gz-cmake/blob/5adbd46a4e8a96f2bc67538bcca2406bb464ace3/cmake/IgnUtils.cmake#L325-L327
```cmake
      foreach(component ${ign_find_package_PRIVATE_FOR})
        set(${component}_${PACKAGE_NAME}_PRIVATE true)
      endforeach()
```
https://github.com/gazebosim/gz-cmake/blob/5adbd46a4e8a96f2bc67538bcca2406bb464ace3/cmake/IgnUtils.cmake#L332-L336
```cmake
      foreach(component ${ign_find_package_REQUIRED_BY})
        if(NOT ${component}_${PACKAGE_NAME}_PRIVATE)
          ign_string_append(${component}_CMAKE_DEPENDENCIES "${${PACKAGE_NAME}_find_dependency}" DELIM "\n")
        endif()
      endforeach()
```
https://github.com/gazebosim/gz-cmake/blob/5adbd46a4e8a96f2bc67538bcca2406bb464ace3/cmake/IgnUtils.cmake#L433-L439
```cmake
            if(${component}_${PACKAGE_NAME}_PRIVATE)
              # If this is a private library or module, use the _PRIVATE suffix
              set(${component}_${PACKAGE_NAME}_PKGCONFIG_TYPE ${component}_${${PACKAGE_NAME}_PKGCONFIG_TYPE}_PRIVATE)
            else()
              # Otherwise, use the plain type
              set(${component}_${PACKAGE_NAME}_PKGCONFIG_TYPE ${component}_${${PACKAGE_NAME}_PKGCONFIG_TYPE})
            endif()
```
https://github.com/gazebosim/gz-cmake/blob/5adbd46a4e8a96f2bc67538bcca2406bb464ace3/cmake/IgnUtils.cmake#L433-L443
```cmake
            if(${component}_${PACKAGE_NAME}_PRIVATE)
              # If this is a private library or module, use the _PRIVATE suffix
              set(${component}_${PACKAGE_NAME}_PKGCONFIG_TYPE ${component}_${${PACKAGE_NAME}_PKGCONFIG_TYPE}_PRIVATE)
            else()
              # Otherwise, use the plain type
              set(${component}_${PACKAGE_NAME}_PKGCONFIG_TYPE ${component}_${${PACKAGE_NAME}_PKGCONFIG_TYPE})
            endif()

            # Append the entry as a string onto the component-specific variable
            # for whichever required type we selected
            ign_string_append(${${component}_${PACKAGE_NAME}_PKGCONFIG_TYPE} ${${PACKAGE_NAME}_PKGCONFIG_ENTRY})
```
https://github.com/gazebosim/gz-cmake/blob/ign-cmake2/cmake/ignition-component-config.cmake.in#L80
```cmake
@component_cmake_dependencies@
```

In fact, the configuration of cmake should be the same as that of pkgconfig. So I think "make PRIVATE_FOR not suppress find_dependency" is consistent with the modification (current modification) of directly removing `PRIVATE_FOR` from these ports that should be exported.

**Comment by dg0yt** [comment]
> In fact, the configuration of cmake should be the same as that of pkgconfig. So I think "make PRIVATE_FOR not suppress find_dependency" is consistent with the modification (current modification) of directly removing `PRIVATE_FOR` from these ports that should be exported.

I cannot test it now but I disagree. I expect that your current modification changes the link type from `PRIVATE` to `PUBLIC`. This would be notable in the generated config by losing the `$<LINK_ONLY:...>` wrapping in `INTERFACE_LINK_LIBRARIES` properties, having side effects on consumers (e.g. adding definitions and include paths).
That's the whole point: You need a `find_dependency` for every library (which is not header-only), to satisfy static linkage.

**Comment by JackBoosY** [comment]
Already tested usage on x64-linux.

**Pull request merged**